File: sdk/dart/lib/src/scheduler.dart
/// Background check scheduling (SPEC.md section 12.2 companion).
///
/// Throttling already lives in [UpdatePolicy.shouldCheck] / [RupUpdater.check].
/// This type only wakes on a timer, calls `check(force: false)`, and reschedules
/// from the result — so hosts do not re-implement "wait 24h / 1h" themselves.
library;

import 'dart:async';

import 'state.dart';
import 'updater.dart';

/// Runs throttled update checks on start and on a wake timer.
///
/// The scheduler never bypasses throttling. User-initiated checks stay on
/// [RupUpdater.check] with `force: true` outside this type.
class UpdateScheduler {
  UpdateScheduler({
    required this.check,
    required this.onResult,
    this.policy = const UpdatePolicy(),
    this.log,
    this.now,
    this.scheduleTimer,
  });

  /// Usually `updater.check` or a host wrapper that adds logging / support gates.
  final Future<UpdateCheckResult> Function({bool force}) check;

  /// Called for every completed check except [CheckThrottled] (those only
  /// reschedule). Hosts typically open UI on [UpdateAvailable].
  final void Function(UpdateCheckResult result) onResult;

  final UpdatePolicy policy;
  final void Function(String message)? log;

  /// Injectable clock for tests.
  final DateTime Function()? now;

  /// Injectable timer factory for tests. Defaults to [Timer].
  final Timer Function(Duration delay, void Function() callback)? scheduleTimer;

  bool _running = false;
  bool _inflight = false;
  Timer? _timer;
  int _generation = 0;

  bool get isRunning => _running;

  /// Starts the loop. When [checkOnStart] is true (default), a tick runs
  /// immediately; otherwise the first wake is [UpdatePolicy.afterFailure]
  /// (short) so a newly started host still converges quickly after failures.
  void start({bool checkOnStart = true}) {
    if (_running) return;
    _running = true;
    _generation++;
    log?.call('update scheduler started (checkOnStart=$checkOnStart)');
    if (checkOnStart) {
      unawaited(_tick());
    } else {
      _arm(policy.afterFailure);
    }
  }

  /// Cancels the pending wake. An in-flight check may still finish; its result
  /// is ignored for rescheduling once stopped.
  void stop() {
    if (!_running) return;
    _running = false;
    _generation++;
    _timer?.cancel();
    _timer = null;
    log?.call('update scheduler stopped');
  }

  DateTime _now() => now?.call() ?? DateTime.now();

  void _arm(Duration delay) {
    _timer?.cancel();
    if (!_running) return;
    final safe = delay.isNegative ? Duration.zero : delay;
    final gen = _generation;
    final create = scheduleTimer ?? Timer.new;
    _timer = create(safe, () {
      if (!_running || gen != _generation) return;
      unawaited(_tick());
    });
  }

  void _armUntil(DateTime when) {
    final wait = when.difference(_now());
    _arm(wait < Duration.zero ? Duration.zero : wait);
  }

  Future<void> _tick() async {
    if (!_running || _inflight) return;
    _inflight = true;
    final gen = _generation;
    try {
      final result = await check(force: false);
      if (!_running || gen != _generation) return;

      switch (result) {
        case CheckThrottled(:final nextAllowedAt):
          log?.call('scheduler throttled until $nextAllowedAt');
          _armUntil(nextAllowedAt);
        case UpToDate():
          onResult(result);
          _arm(policy.afterSuccess);
        case UpdateAvailable():
          onResult(result);
          // Surface to the user, but do not poll the network again until the
          // success interval elapses (install / skip happens on the host).
          _arm(policy.afterSuccess);
        case FallbackRequired():
          onResult(result);
          _arm(policy.afterSuccess);
        case CheckFailed():
          onResult(result);
          _arm(policy.afterFailure);
      }
    } catch (error) {
      log?.call('scheduler tick failed: $error');
      if (_running && gen == _generation) {
        _arm(policy.afterFailure);
      }
    } finally {
      _inflight = false;
    }
  }
}

File: sdk/node/src/scheduler.ts
/**
 * Background check scheduling (SPEC.md section 12.2 companion).
 *
 * Throttling lives in `shouldCheck` / `RupUpdater.check`. Periodic ticks call
 * `check({ force: false })`. The startup tick may call `check({ force: true })`
 * when `forceOnStart` is true.
 */

import { defaultUpdatePolicy, type UpdatePolicy } from "./state.js";
import type { UpdateCheckResult } from "./updater.js";

/** Scheduler and throttling knobs for a running client. */
export interface UpdateRuntimeConfig {
  /** Whether `start()` runs an immediate tick. */
  checkOnStart: boolean;
  /** When true, the startup tick ignores throttling. */
  forceOnStart: boolean;
  policy: UpdatePolicy;
}

export const defaultRuntimeConfig: UpdateRuntimeConfig = {
  checkOnStart: true,
  forceOnStart: true,
  policy: defaultUpdatePolicy,
};

/** Parses host JSON. Unknown keys are ignored. */
export function runtimeConfigFromJson(raw: unknown): UpdateRuntimeConfig {
  const json =
    typeof raw === "object" && raw !== null
      ? (raw as Record<string, unknown>)
      : {};
  const hours = (value: unknown, fallbackMs: number): number =>
    typeof value === "number" ? value * 60 * 60 * 1000 : fallbackMs;

  return {
    checkOnStart:
      typeof json.checkOnStart === "boolean" ? json.checkOnStart : true,
    forceOnStart:
      typeof json.forceOnStart === "boolean" ? json.forceOnStart : true,
    policy: {
      ...defaultUpdatePolicy,
      afterSuccessMs: hours(
        json.afterSuccessHours,
        defaultUpdatePolicy.afterSuccessMs,
      ),
      afterFailureMs: hours(
        json.afterFailureHours,
        defaultUpdatePolicy.afterFailureMs,
      ),
    },
  };
}

export interface UpdateSchedulerOptions {
  /** Usually `updater.check` or a host wrapper that adds logging / gates. */
  check: (options: { force: boolean }) => Promise<UpdateCheckResult>;
  /**
   * Called for every completed check except `check-throttled` (those only
   * reschedule). Hosts typically open UI on `update-available`.
   */
  onResult: (result: UpdateCheckResult) => void;
  runtime?: Partial<UpdateRuntimeConfig>;
  log?: (message: string) => void;
  now?: () => Date;
}

/** Runs throttled update checks on start and on a wake timer. */
export class UpdateScheduler {
  private readonly runtime: UpdateRuntimeConfig;
  private readonly check: UpdateSchedulerOptions["check"];
  private readonly onResult: UpdateSchedulerOptions["onResult"];
  private readonly log?: (message: string) => void;
  private readonly nowFn: () => Date;

  private running = false;
  private inflight = false;
  private timer: NodeJS.Timeout | null = null;
  private generation = 0;

  constructor(options: UpdateSchedulerOptions) {
    this.check = options.check;
    this.onResult = options.onResult;
    if (options.log) this.log = options.log;
    this.nowFn = options.now ?? (() => new Date());
    this.runtime = {
      ...defaultRuntimeConfig,
      ...options.runtime,
      policy: { ...defaultUpdatePolicy, ...options.runtime?.policy },
    };
  }

  get policy(): UpdatePolicy {
    return this.runtime.policy;
  }

  get isRunning(): boolean {
    return this.running;
  }

  start(overrides: { checkOnStart?: boolean; forceOnStart?: boolean } = {}): void {
    if (this.running) return;
    this.running = true;
    this.generation++;

    const onStart = overrides.checkOnStart ?? this.runtime.checkOnStart;
    const force = overrides.forceOnStart ?? this.runtime.forceOnStart;
    this.log?.(
      `update scheduler started (checkOnStart=${onStart} forceOnStart=${force})`,
    );

    if (onStart) {
      void this.tick(force);
    } else {
      this.arm(this.policy.afterFailureMs);
    }
  }

  /**
   * Cancels the pending wake. An in-flight check may still finish; its result is
   * ignored for rescheduling once stopped.
   */
  stop(): void {
    if (!this.running) return;
    this.running = false;
    this.generation++;
    if (this.timer !== null) {
      clearTimeout(this.timer);
      this.timer = null;
    }
    this.log?.("update scheduler stopped");
  }

  private arm(delayMs: number): void {
    if (this.timer !== null) clearTimeout(this.timer);
    this.timer = null;
    if (!this.running) return;
    const safe = Math.max(0, delayMs);
    const generation = this.generation;
    this.timer = setTimeout(() => {
      if (!this.running || generation !== this.generation) return;
      void this.tick(false);
    }, safe);
    this.timer.unref?.();
  }

  private armUntil(when: Date): void {
    this.arm(when.getTime() - this.nowFn().getTime());
  }

  private async tick(force: boolean): Promise<void> {
    if (!this.running || this.inflight) return;
    this.inflight = true;
    const generation = this.generation;
    try {
      const result = await this.check({ force });
      if (!this.running || generation !== this.generation) return;

      switch (result.kind) {
        case "check-throttled":
          this.log?.(`scheduler throttled until ${result.nextAllowedAt.toISOString()}`);
          this.armUntil(result.nextAllowedAt);
          break;
        case "up-to-date":
        case "update-available":
        case "fallback-required":
          // Surface to the user, but do not poll the network again until the
          // success interval elapses (install / skip happens on the host).
          this.onResult(result);
          this.arm(this.policy.afterSuccessMs);
          break;
        case "check-failed":
          this.onResult(result);
          this.arm(this.policy.afterFailureMs);
          break;
      }
    } catch (error) {
      this.log?.(`scheduler tick failed: ${String(error)}`);
      if (this.running && generation === this.generation) {
        this.arm(this.policy.afterFailureMs);
      }
    } finally {
      this.inflight = false;
    }
  }
}
